pkg/cluster-operator/customcluster_controller_test.go

```diff
@@ -0,0 +1,485 @@
+			},
+		},
+	}
+	testKcp := &controlplanev1.KubeadmControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-kcp",
+			Finalizers: []string{
+				"Kcp-finalizer",
+			},
+		},
+	}
+	r := &CustomClusterController{
+		Client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).
+			WithObjects(testCustomCluster, testCustomMachine, testKcp).Build(),
+	}
+
+	err := r.deleteResource(ctx, testCustomCluster, testCustomMachine, testKcp)
+	if err != nil {
+		t.Errorf("customcluster_controller deleteResource() error is %v", err)
+		return
+	}
+}
+
+func TestEnsureFinalizerAndOwnerRef(t *testing.T) {
+	ctx := context.Background()
+	testCustomCluster := &v1alpha1.CustomCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-customcluster",
+		},
+	}
+	testCustomMachine := &v1alpha1.CustomMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-machine",
+		},
+	}
+	testKcp := &controlplanev1.KubeadmControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-kcp",
+		},
+	}
+	testClusterHosts := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-clusterhost",
+		},
+	}
+	testClusterConfig := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-clusterconfig",
+		},
+	}
+
+	r := &CustomClusterController{
+		Client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).
+			WithRuntimeObjects(testCustomCluster, testCustomMachine, testKcp, testClusterConfig, testClusterHosts).
+			Build(),
+	}
+	err := r.ensureFinalizerAndOwnerRef(ctx, testClusterHosts, testClusterConfig, testCustomCluster, testCustomMachine, testKcp)
+	if err != nil {
+		t.Errorf("customcluster_controller ensureFinalizerAndOwnerRef() error is %v", err)
+		return
+	}
+}
+
+func TestReconcileDelete(t *testing.T) {
+	ctx := context.Background()
+	testCustomCluster := &v1alpha1.CustomCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-customcluster",
+			UID:       "customcluster",
+		},
+	}
+	testCustomMachine := &v1alpha1.CustomMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-machine",
+		},
+	}
+	testKcp := &controlplanev1.KubeadmControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-kcp",
+		},
+	}
+	testPod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "test-pod",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Kind: CustomClusterKind,
+					Name: "test-customcluster",
+					UID:  "customcluster",
+				},
+			},
+		},
+	}
+	r := &CustomClusterController{
+		Client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).
+			WithRuntimeObjects(testCustomCluster, testCustomMachine, testKcp, testPod).
+			Build(),
+	}
+
+	result, err := r.reconcileDelete(ctx, testCustomCluster, testCustomMachine, testKcp)
```

Comment:
You can construct more cases to coverage more branches, go or goland can test the code coverage